pkg/asset-manager-utils/test/RebalancingRelayer.test.ts

```diff
@@ -0,0 +1,234 @@
+import { BigNumberish, fp } from '@balancer-labs/v2-helpers/src/numbers';
+import {
+  encodeExitWeightedPool,
+  encodeJoinWeightedPool,
+} from '@balancer-labs/v2-helpers/src/models/pools/weighted/encoding';
+
+describe.only('RebalancingRelayer', function () {
+  let user: SignerWithAddress, admin: SignerWithAddress, tokens: TokenList, poolId: string;
+  let vault: Contract, authorizer: Contract, relayer: Contract, pool: Contract, assetManager: Contract;
+
+  before('setup signer', async () => {
+    [, admin, user] = await ethers.getSigners();
+  });
+
+  sharedBeforeEach('deploy relayer', async () => {
+    authorizer = await deploy('v2-vault/Authorizer', { args: [admin.address] });
+    vault = await deploy('v2-vault/Vault', { args: [authorizer.address, ZERO_ADDRESS, 0, 0] });
+    relayer = await deploy('RebalancingRelayer', { args: [vault.address] });
+  });
+
+  sharedBeforeEach('deploy sample pool', async () => {
+    tokens = await TokenList.create(['DAI', 'MKR'], { sorted: true });
+    assetManager = await deploy('MockAssetManager');
+    pool = await deploy('v2-pool-utils/MockRelayedBasePool', {
+      args: [
+        vault.address,
+        GeneralPool,
+        'BPT',
+        'BPT',
+        tokens.addresses,
+        Array(tokens.length).fill(assetManager.address),
+        fp(0.1),
+        0,
+        0,
+        relayer.address,
+        ZERO_ADDRESS,
+      ],
+    });
+    poolId = await pool.getPoolId();
+  });
+
+  describe('vault', () => {
+    it('uses the given vault', async () => {
+      expect(await relayer.vault()).to.be.equal(vault.address);
+    });
+  });
+
+  describe('join', () => {
+    let request: { assets: string[]; maxAmountsIn: BigNumberish[]; userData: string; fromInternalBalance: boolean };
+
+    sharedBeforeEach('build join request', async () => {
+      const amountsIn = Array(tokens.length).fill(fp(10));
+      request = {
+        assets: tokens.addresses,
+        maxAmountsIn: amountsIn,
+        userData: encodeJoinWeightedPool({ kind: 'Init', amountsIn }),
+        fromInternalBalance: false,
+      };
+    });
+
+    context('when going through the relayer', () => {
+      context('when the relayer is allowed to join', () => {
+        sharedBeforeEach('allow relayer', async () => {
+          const action = await actionId(vault, 'joinPool');
+          await authorizer.connect(admin).grantRole(action, relayer.address);
+        });
+
+        context('when the user did allow the relayer', () => {
+          sharedBeforeEach('allow relayer', async () => {
+            await vault.connect(user).setRelayerApproval(user.address, relayer.address, true);
+          });
+
+          it('joins the pool', async () => {
+            const previousUserBalance = await pool.balanceOf(user.address);
+            const previousRelayerBalance = await pool.balanceOf(relayer.address);
+
+            await relayer.connect(user).joinPool(poolId, request);
```

Comment:
What do you think about emitting an event in the pool with the join arguments it received? Overkill?